runtime/runtime.go

```diff
@@ -0,0 +1,132 @@
+package runtime
+
+import (
+	"fmt"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/runtime/protoiface"
+	"io"
+	"math/bits"
+)
+
+func Sov(x uint64) (n int) {
+	return (bits.Len64(x|1) + 6) / 7
+}
+func Soz(x uint64) (n int) {
+	return Sov((x << 1) ^ uint64(int64(x)>>63))
+}
+
+func EncodeVarint(dAtA []byte, offset int, v uint64) int {
+	offset -= Sov(v)
+	base := offset
+	for v >= 1<<7 {
+		dAtA[offset] = uint8(v&0x7f | 0x80)
+		v >>= 7
+		offset++
+	}
+	dAtA[offset] = uint8(v)
+	return base
+}
+
+func Skip(dAtA []byte) (n int, err error) {
+	l := len(dAtA)
+	iNdEx := 0
+	depth := 0
+	for iNdEx < l {
+		var wire uint64
+		for shift := uint(0); ; shift += 7 {
+			if shift >= 64 {
+				return 0, ErrIntOverflow
+			}
+			if iNdEx >= l {
+				return 0, io.ErrUnexpectedEOF
+			}
+			b := dAtA[iNdEx]
+			iNdEx++
+			wire |= (uint64(b) & 0x7F) << shift
+			if b < 0x80 {
+				break
+			}
+		}
+		wireType := int(wire & 0x7)
+		switch wireType {
+		case 0:
+			for shift := uint(0); ; shift += 7 {
+				if shift >= 64 {
+					return 0, ErrIntOverflow
+				}
+				if iNdEx >= l {
+					return 0, io.ErrUnexpectedEOF
+				}
+				iNdEx++
+				if dAtA[iNdEx-1] < 0x80 {
+					break
+				}
+			}
+		case 1:
+			iNdEx += 8
+		case 2:
+			var length int
+			for shift := uint(0); ; shift += 7 {
+				if shift >= 64 {
+					return 0, ErrIntOverflow
+				}
+				if iNdEx >= l {
+					return 0, io.ErrUnexpectedEOF
+				}
+				b := dAtA[iNdEx]
+				iNdEx++
+				length |= (int(b) & 0x7F) << shift
+				if b < 0x80 {
+					break
+				}
+			}
+			if length < 0 {
+				return 0, ErrInvalidLength
+			}
+			iNdEx += length
+		case 3:
+			depth++
+		case 4:
+			if depth == 0 {
+				return 0, ErrUnexpectedEndOfGroup
+			}
+			depth--
+		case 5:
+			iNdEx += 4
+		default:
+			return 0, fmt.Errorf("proto: illegal wireType %d", wireType)
+		}
+		if iNdEx < 0 {
+			return 0, ErrInvalidLength
+		}
+		if depth == 0 {
+			return iNdEx, nil
+		}
+	}
+	return 0, io.ErrUnexpectedEOF
+}
+
+func MarshalFlagsToOptions(flags protoiface.MarshalInputFlags) proto.MarshalOptions {
+	return proto.MarshalOptions{
+		NoUnkeyedLiterals: struct{}{},
+		AllowPartial:      false,
+		Deterministic:     true,
+		UseCachedSize:     flags == protoiface.MarshalUseCachedSize,
+	}
+}
+
+func UnmarshalInputToOptions(input protoiface.UnmarshalInput) proto.UnmarshalOptions {
+	return proto.UnmarshalOptions{
+		NoUnkeyedLiterals: struct{}{},
+		Merge:             false,
+		AllowPartial:      false,
+		DiscardUnknown:    input.Flags == protoiface.UnmarshalDiscardUnknown,
```

Comment:
This check should be done with a bitwise operation, refer to os.OpenFile, but it should be something like ```input.Flags&protoiface.UnmarshalDiscardUnknown==0```